core/kernel/api/class.ModelFactory.php

```diff
@@ -16,112 +17,143 @@
                 }
             }
         }
-        
-        return true;
+
+        return $returnValue;
     }
-    
+
     /**
      * Adds a statement to the ontology if it does not exist yet
-     * 
+     *
      * @author "Joel Bout, <joel@taotesting.com>"
-     * @param int $modelId
+     *
+     * @param int    $modelId
      * @param string $subject
      * @param string $predicate
      * @param string $object
      * @param string $lang
+     * @param string $author
+     *
+     * @return bool Was a row added?
      */
-    private function addStatement($modelId, $subject, $predicate, $object, $lang = null) {
-        $result = core_kernel_classes_DbWrapper::singleton()->query(
+    public function addStatement($modelId, $subject, $predicate, $object, $lang = null, $author = self::DEFAULT_AUTHOR)
+    {
+        // TODO: refactor this to use a triple store abstraction.
+        $result = $this->dbWrapper->query(
             'SELECT count(*) FROM statements WHERE modelid = ? AND subject = ? AND predicate = ? AND object = ? AND l_language = ?',
-            array($modelId, $subject, $predicate, $object, (is_null($lang)) ? '' : $lang)
+            [$modelId, $subject, $predicate, $object, (is_null($lang)) ? '' : $lang]
         );
-        
-        if (intval($result->fetchColumn()) === 0) {
-            $dbWrapper = core_kernel_classes_DbWrapper::singleton();
-            $date = $dbWrapper->getPlatForm()->getNowExpression();
-
-            $dbWrapper->insert(
-                'statements',
-                array(
-                    'modelid' =>  $modelId,
-                    'subject' =>$subject,
-                    'predicate'=> $predicate,
-                    'object' => $object,
-                    'l_language' => is_null($lang) ? '' : $lang,
-                    'author' => 'http://www.tao.lu/Ontologies/TAO.rdf#installator',
-                    'epoch' => $date
-                )
-            );
+
+        if (intval($result->fetchColumn()) > 0) {
+            return false;
         }
+
+        $date = $this->dbWrapper->getPlatForm()->getNowExpression();
+
+        return (bool) $this->dbWrapper->insert(
+            'statements',
+            [
+                'id' => $this->getUniquePrimaryKey(),
+                'modelid' => $modelId,
+                'subject' => $subject,
+                'predicate' => $predicate,
+                'object' => $object,
+                'l_language' => is_null($lang) ? '' : $lang,
+                'author' => $author,
+                'epoch' => $date,
+            ]
+        );
     }
-    
 
-}
\ No newline at end of file
+    /**
+     * Generates a unique primary key for gcp Spanner sharding.
+     *
+     * @TODO: confirm this is the optimal solution for gcp Spanner.
+     *
+     * @return string
+     */
+    public function getUniquePrimaryKey()
```

Comment:
https://www.php.net/manual/en/function.uniqid.php#94959
```
public static function v4() {
    return sprintf('%04x%04x-%04x-%04x-%04x-%04x%04x%04x',

      // 32 bits for "time_low"
      mt_rand(0, 0xffff), mt_rand(0, 0xffff),

      // 16 bits for "time_mid"
      mt_rand(0, 0xffff),

      // 16 bits for "time_hi_and_version",
      // four most significant bits holds version number 4
      mt_rand(0, 0x0fff) | 0x4000,

      // 16 bits, 8 bits for "clk_seq_hi_res",
      // 8 bits for "clk_seq_low",
      // two most significant bits holds zero and one for variant DCE1.1
      mt_rand(0, 0x3fff) | 0x8000,

      // 48 bits for "node"
      mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0xffff)
    );
  }
```
link : https://cloud.google.com/spanner/docs/schema-and-data-model
@ Choosing a primary key
.............................................................
Use a Universally Unique Identifier (UUID). Version 4 UUID is recommended because it uses random values in the high-order bits. Don't use a UUID algorithm (such as version 1 UUID) that stores the timestamp in the high order bits.

Reply:
In all cases, you should be careful not to create hotspots with the choice of your primary key. For example, if you insert records with a monotonically increasing integer as the key, you'll always insert at the end of your key space. This is undesirable because Cloud Spanner divides data among servers by key ranges, which means your inserts will be directed at a single server, creating a hotspot.

Reply:
please while testing don't forget to provide more attention to this issue :) 
I think some benchmarks will give us more details if the chosen strategy is ok :)